src/printer.js

```diff
@@ -2048,71 +2050,198 @@ function printMemberChain(node, options, print) {
+// is left that way.
+// Not only that, any user-written-line containing multiple JSX siblings
+// should also be kept on one line if possible,
+// so each user-written-line is wrapped in its own group.
+//
+// Elements that contain newlines or don't fit on a single line (recursively)
+// are fully-split, using hardline and shouldBreak: true.
+//
+// To support that case properly, all leading and trailing spaces
+// are stripped from the list of children, and replaced with a single hardline.
 function printJSXElement(path, options, print) {
   const n = path.getValue();
-  var openingLines = path.call(print, "openingElement");
 
-  let elem;
+  // turn <div></div> into <div />
+  if (isEmptyJSXElement(n)) {
+    n.openingElement.selfClosing = true;
+    delete n.closingElement;
+  }
+
+  // if no children, just print the opening element
+  const openingLines = path.call(print, "openingElement");
   if (n.openingElement.selfClosing) {
     assert.ok(!n.closingElement);
+    return openingLines;
+  }
 
-    elem = openingLines;
-  } else {
-    var children = [];
+  const children = printJSXChildren(path, options, print);
+  let hasAnyHardLine = false;
+
+  // trim trailing whitespace, recording if there was a hardline
+  while (children.length && util.getLast(children).type === "line") {
+    if (hasHardLine(util.getLast(children))) hasAnyHardLine = true;
+    children.pop();
+  }
 
-    path.map(
-      function(childPath) {
-        var child = childPath.getValue();
+  // trim leading whitespace, recording if there was a hardline
+  while (children.length && children[0].type === "line") {
+    if (hasHardLine(children[0])) hasAnyHardLine = true;
+    children.shift();
+  }
 
-        if (
-          namedTypes.Literal.check(child) && typeof child.value === "string"
-        ) {
-          if (/\S/.test(child.value)) {
-            const beginBreak = child.value.match(/^\s*\n/);
-            const endBreak = child.value.match(/\n\s*$/);
-
-            children.push(
-              beginBreak ? hardline : "",
-              child.value.replace(/^\s+|\s+$/g, endBreak ? "" : " "),
-              endBreak ? hardline : ""
-            );
-          } else if (/\n/.test(child.value)) {
-            children.push(hardline);
-          }
-        } else {
-          children.push(print(childPath));
-        }
-      },
-      "children"
-    );
+  // group by line, recording if there was a hardline.
+  let childrenGroupedByLine;
+  if (children.length === 1) {
+    if (!hasAnyHardLine && hasHardLine(children[0])) hasAnyHardLine = true;
 
-    var mostChildren = children.slice(0, -1);
-    var closingLines = path.call(print, "closingElement");
+    // no need for groups, and a lone jsx whitespace doesn't work otherwise
+    childrenGroupedByLine = [ hardline, children[0] ];
+  } else {
+    //  prefill leading newline, and initialize the first line's group
+    childrenGroupedByLine = [ hardline, group(concat([])) ];
```

Comment:
@jlongster you'd asked about this before... still one of the uglier bits in this PR IMO, but the only prettier ways I can think of would sacrifice perf. look ok to you?